### PR TITLE
feat(ui-react-native): Radio

### PR DESCRIPTION
**Pull request by joebuono**
#### Description of changes

Updated PR in response to feedback. 

In order to allow the `size` prop to accept a `number` or a string union of `'small' | 'medium' | 'large'`, I wrote a few utility functions in `getRadioButtonStyles`. 

https://user-images.githubusercontent.com/48109584/192354990-ef5691a2-8704-445c-95d0-be734c6ecf25.mov

#### Description of how you validated changes

Ran locally in both iOS and Android. 

#### Checklist

- [X] PR description included
- [ ] `yarn test` passes
- [X] Tests are updated

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: examples/react-native/src/features/InAppMessaging/Demo/Demo.tsx
Comment (joebuono) [inline examples/react-native/src/features/InAppMessaging/Demo/Demo.tsx:29]: I changed the type for the `size` prop to match the [RWA](https://ui.docs.amplify.aws/react/components/radiogroupfield#sizes).

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (joebuono) [inline packages/react-native/src/primitives/Radio/Radio.tsx:54]: I still need to handle platform-specific touch target sizes for `hitSlop`

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (joebuono) [inline packages/react-native/src/primitives/Radio/styles.ts:49]: I'm not sure what our naming convention is for different states.

File: packages/react-native/src/primitives/Radio/types.ts
Comment (joebuono) [inline packages/react-native/src/primitives/Radio/types.ts:39]: Totally open to suggestions about how to type this better.

**Comment by ioanabrooks** [comment]
Leaving a general note, I see the demo videos posted are iOS only, want to make sure that this is tested in android as well, to help uncover any potential issues early. By running `yarn run react-native-example storybook` in one terminal and `yarn run react-native-example dev` in a second one, you can have both simulators open and control stories through the storybook UI.

For styling related questions, I would defer to later when we have theming in place, we will go back and change styling related things when adding theming. But this is a great start!

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (joebuono) [inline packages/react-native/src/primitives/Radio/styles.ts:22]: I'm wondering if it'd be easier to just set an arbitrarily large `borderRadius` in the main `radio` so that it is always a circle regardless of its size. Same goes for `radioButton`.

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/Radio.tsx:69]: Maybe we can just have the `label` logic above and let the parent `flexDirection` value control the layout flow? For example when the `label` should render leading the Radio icon using `column-reverse`

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/Radio.tsx:28]: Can we consolidate all the container style logic here and rename as `containerStyle`?

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/styles.ts:12]: `30` is really large for a default icon on mobile, maybe we can update the size base values to:
- small: 16
- medium: 20
- large: 24

Ultimately the theme will control but think these are sensible defaults. Thoughts?

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/Radio.tsx:55]: Do we need both a `View` and a `Pressable`? Thinking we can just add the `style` on the `Pressable`?

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/Radio.tsx:16]: Just a reminder that we need to make sure we have logic in place to handle both controlled and uncontrolled use cases ([React docs](https://reactjs.org/docs/uncontrolled-components.html))

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/Radio.tsx:4]: Should be able to leave off the `/index`

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/Radio.tsx:7]: Nit: swap the order of these, `./styles` before `./types`

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/styles.ts:14]: We want to avoid using odd numbers for assigning space related values, they prevent centering of layout items when used in theme structures. Our default spacing size will be `8`, and the rest of the values will be derived based on that. Suggest using `4` here

File: packages/react-native/src/primitives/Radio/types.ts
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/types.ts:6]: Do we need this comment?

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/Radio.tsx:61]: Should this be `buttonStyle`?

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/Radio.tsx:42]: If we follow the UX of our React web `RadioGroupField` implementation, the entire `Radio` element should be handle press, so let's replace this `View` with a `Pressable`

**Review by calebpollman — COMMENTED** [review]
Great start! Let me know if you need clarification on any of the feedback

File: examples/react-native/storybook/stories/Radio.stories.tsx
Comment (ioanabrooks) [inline examples/react-native/storybook/stories/Radio.stories.tsx:20]: Nice coverage of scenarios, stealing this for checkbox :)

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (joebuono) [reply packages/react-native/src/primitives/Radio/Radio.tsx:20]: As in [something like this](https://github.com/aws-amplify/amplify-ui/blob/in-app-messaging/main/packages/react-native/src/primitives/Checkbox/Checkbox.tsx#L30)?

File: packages/react-native/src/primitives/Radio/types.ts
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/types.ts:15]: Since these aren't really "button" styles, can we use names that are more indicative of their usage? Example:
```suggestion
  radioContainerStyle?: StyleProp<ViewStyle>;
  radioDotStyle?: StyleProp<ViewStyle>;
```

File: packages/react-native/src/primitives/Radio/types.ts
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/types.ts:8]: Should `RadioProps` extend from `PressableProps`?

File: packages/react-native/src/primitives/Radio/types.ts
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/types.ts:6]: Union type naming should be singular form. Also let's add `number` to the union instead of declaring another union on ln 16

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/styles.ts:27]: What do you think of removing these "size" related objects and returning the style objects from the dimension creation util?

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/Radio.tsx:47]: Do we always want this to be `5`? What are the expected touch target sizes for the platforms?

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/Radio.tsx:4]: To avoid potential circular dependencies, we should import directly from the `Label` directory

**Review by calebpollman — COMMENTED** [review]
Getting closer, thanks for making the updates!

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (joebuono) [reply packages/react-native/src/primitives/Radio/styles.ts:28]: IMO, that would make the outlines a bit too thin:
Screen Shot 2022-09-26 at 5 35 14 PM

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (joebuono) [reply packages/react-native/src/primitives/Radio/styles.ts:15]: I just copied the colors from the [React web Radio](https://ui.docs.amplify.aws/react/components/radiogroupfield), which has different colors for the border and the button:
Screen Shot 2022-09-26 at 5 41 52 PM

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (calebpollman) [reply packages/react-native/src/primitives/Radio/styles.ts]: Can we update these sizes?

File: packages/react-native/src/primitives/Radio/types.ts
Comment (joebuono) [reply packages/react-native/src/primitives/Radio/types.ts:6]: Ok. Since we are likely to have a `size` prop for other primitives as well, I think we'll eventually want to find another place to define the `Sizes` type.

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (calebpollman) [reply packages/react-native/src/primitives/Radio/styles.ts:15]: Not certain why that was done for web. Let's go ahead and align them to the same color

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (joebuono) [reply packages/react-native/src/primitives/Radio/styles.ts:27]: We could do that, but I'm wondering how customers would then be able to override the styles for a particular size (e.g., `large`).

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (calebpollman) [reply packages/react-native/src/primitives/Radio/styles.ts:28]: Maybe a default size of `1` is appropriate

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (calebpollman) [reply packages/react-native/src/primitives/Radio/styles.ts:27]: They would just pass the value they want in the `size` prop?

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (joebuono) [reply packages/react-native/src/primitives/Radio/Radio.tsx:62]: Looks like the minimum touch target size for [iOS](https://www.w3.org/WAI/GL/mobile-a11y-tf/wiki/M2#:~:text=iOS%20Human%20Interface%20Guidelines%3A%20Adaptivity,pixels%20wide%2044%20pixels%20tall.) is 44px wide and 44px tall. For [Android](https://www.w3.org/WAI/GL/mobile-a11y-tf/wiki/M2#:~:text=Android%20Developers%3A%20Metrics%20and%20Grids%20recommends%20targets%20be%20at%20least%2048dp%20high%20and%20wide%20so%20that%20they%20will%20never%20be%20smaller%20than%20a%20minimum%20recommended%20target%20size%20of%207mm.) it's 48dp X 48dp. 

In addition to taking the platform into consideration, do we want to dynamically calculate `hitSlop` based on the size of the element? For example, if the Radio is 12 high and 40 wide, then we would want the hitSlop to be something like:

```
hitSlop={{ top: 16, right: 2, bottom: 16, left: 2 }}
```
in order to increase the target size to the minimum. 

However, if the element is larger (e.g., 250 X 250), then adding `hitSlop` wouldn't be necessary. 

Thus I'm wondering, if `hitSlop` is something that we want to handle for all of our touchable/pressable primitives, perhaps we should consider writing a util for it? The util could take both the platform and the size of the element, and return the calculated `hitSlop`.

Thoughts?

File: examples/react-native/storybook/stories/Radio.stories.tsx
Comment (ioanabrooks) [inline examples/react-native/storybook/stories/Radio.stories.tsx:12]: This is no longer needed.

File: examples/react-native/storybook/stories/Radio.stories.tsx
Comment (ioanabrooks) [inline examples/react-native/storybook/stories/Radio.stories.tsx:20]: If you wanted to, you could use storybook's addon -> knobs:
```
 <Radio labelPosition={select(
        'LabelPosition',
        ['start', 'end', 'top', 'bottom'],
        'end'
      )} />
```

File: packages/react-native/src/primitives/Radio/types.ts
Comment (ioanabrooks) [inline packages/react-native/src/primitives/Radio/types.ts:5]: This type could be shared among other components.

File: packages/react-native/src/primitives/Radio/types.ts
Comment (ioanabrooks) [inline packages/react-native/src/primitives/Radio/types.ts:15]: Size type already has number

Comment (joebuono) [reply packages/react-native/src/primitives/Radio/types.ts:15]: Good catch

File: packages/react-native/src/primitives/Radio/types.ts
Comment (joebuono) [reply packages/react-native/src/primitives/Radio/types.ts:5]: I agree. Is there a place we can put shared types?

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/Radio.tsx:54]: Can this be combined?

File: packages/react-native/src/primitives/Radio/types.ts
Comment (calebpollman) [reply packages/react-native/src/primitives/Radio/types.ts:5]: Let's move it in to the Icon types once that PR has been merged

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/Radio.tsx:54]: Can `containerStyle` and `radioStyle` be combined?

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (calebpollman) [reply packages/react-native/src/primitives/Radio/styles.ts:27]: Un-resolving this comment. Lmk if we need to discuss offline

File: packages/react-native/src/primitives/Radio/styles.ts
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/styles.ts:28]: These sizes are really large for mobile, for now let's use:
```
small: 16, 
medium: 20, 
large: 24
```

File: packages/react-native/src/primitives/Radio/Radio.tsx
Comment (joebuono) [reply packages/react-native/src/primitives/Radio/Radio.tsx:54]: Yes

File: packages/react-native/src/primitives/Radio/types.ts
Comment (joebuono) [reply packages/react-native/src/primitives/Radio/types.ts:5]: But will Icon `size` accept `'small' | 'medium' | 'large'`?

File: packages/react-native/src/primitives/Radio/getRadioDimensions.ts
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/getRadioDimensions.ts:35]:
```suggestion
const sizeKey = capitalize(size);
```

File: packages/react-native/src/primitives/Radio/types.ts
Comment (calebpollman) [inline packages/react-native/src/primitives/Radio/types.ts:39]: Thinking we can replace `ViewStyle` with `RadioDimensions` in the return of `getRadioDimensions`:

```suggestion
export type RadioDimensions = Pick<ViewStyle, 'height' | 'width'>;

export interface RadioStyles {
  container: ViewStyle;
  disabled: ViewStyle;
  radioContainer: ViewStyle;
  radioContainerLarge: RadioDimensions;
  radioContainerMedium: RadioDimensions;
  radioContainerSmall: RadioDimensions;
  radioDot: ViewStyle;
  radioDotLarge: RadioDimensions;
  radioDotMedium: RadioDimensions;
  radioDotSmall: RadioDimensions;
}
```

**Review by calebpollman — APPROVED** [review]
Left a couple suggestions but LGTM

**Review by ioanabrooks — APPROVED** [review]
LGTM!

**Comment by calebpollman** [comment]
Please select the "squash and merge" option when merging in the future

**Pull request merged**
